Beam: Linkage Checker gets corrupt JAR file error from amazon-kinesis-client

Investigate whether this kind of corruption is skipped or not.

```
suztomo@suxtomo24:~/beam6$ ./gradlew -Ppublishing -PjavaLinkageArtifactIds=beam-sdks-java-io-kinesis :checkJavaLinkage |tee linkage.log
Configuration on demand is an incubating feature.

> Task :runners:apex:buildDependencyTree
See the report at: file:///usr/local/google/home/suztomo/beam6/runners/apex/build/classes/java/main/org/apache/beam/runners/apex/dependency-tree

> Task :checkJavaLinkage

> Task :checkJavaLinkage FAILED

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.2.1/userguide/command_line_interface.html#sec:command_line_warnings
885 actionable tasks: 156 executed, 729 up-to-date
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Exception in thread "main" java.io.IOException: Corrupt jar file /usr/local/google/home/suztomo/.m2/repository/com/amazonaws/amazon-kinesis-client/1.13.0/amazon-kinesis-client-1.13.0.jar; could not load .unison.com.e007f77498fd27177e2ea931a06dcf50.unison.tmp.amazonaws.services.kinesis.leases.impl.LeaseTaker
        at com.google.cloud.tools.opensource.classpath.ClassDumper.listClasses(ClassDumper.java:393)
        at com.google.cloud.tools.opensource.classpath.ClassDumper.findSymbolReferences(ClassDumper.java:153)
        at com.google.cloud.tools.opensource.classpath.LinkageChecker.create(LinkageChecker.java:78)
        at com.google.cloud.tools.opensource.classpath.LinkageCheckerMain.main(LinkageCheckerMain.java:49)
Caused by: java.lang.ClassNotFoundException: Exception while looking for class .unison.com.e007f77498fd27177e2ea931a06dcf50.unison.tmp.amazonaws.services.kinesis.leases.impl.LeaseTaker: java.io.IOException: Couldn't find: /unison/com/e007f77498fd27177e2ea931a06dcf50/unison/tmp/amazonaws/services/kinesis/leases/impl/LeaseTaker.class
        at org.apache.bcel.util.AbstractClassPathRepository.loadClass(AbstractClassPathRepository.java:77)
        at org.apache.bcel.util.ClassPathRepository.loadClass(ClassPathRepository.java:31)
        at com.google.cloud.tools.opensource.classpath.FixedSizeClassPathRepository.loadClass(FixedSizeClassPathRepository.java:108)
        at com.google.cloud.tools.opensource.classpath.ClassDumper.listClasses(ClassDumper.java:389)
        ... 3 more
Caused by: java.io.IOException: Couldn't find: /unison/com/e007f77498fd27177e2ea931a06dcf50/unison/tmp/amazonaws/services/kinesis/leases/impl/LeaseTaker.class
        at org.apache.bcel.util.ClassPath.getClassFile(ClassPath.java:666)
        at com.google.cloud.tools.opensource.classpath.LinkageCheckClassPath.getInputStream(LinkageCheckClassPath.java:72)
        at org.apache.bcel.util.ClassPath.getInputStream(ClassPath.java:688)
        at org.apache.bcel.util.AbstractClassPathRepository.loadClass(AbstractClassPathRepository.java:75)
        ... 6 more

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':checkJavaLinkage'.
> Process 'command '/usr/lib/jvm/java-8-openjdk-amd64/bin/java'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 8s

```

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
@@ -385,6 +385,9 @@ class ClassDumper {
    */
   private ImmutableSet<JavaClass> listClasses(Path jar) throws IOException {
     ImmutableSet.Builder<JavaClass> javaClasses = ImmutableSet.builder();
+
+    ImmutableList.Builder<String> corruptedClassFileNames = ImmutableList.builder();
+
     for (String classFileName : listClassFileNames(jar)) {
       if (classFileName.startsWith("META-INF.versions.")) {
         // Linkage Checker does not support multi-release JAR (for Java 9+) yet
@@ -402,7 +405,7 @@ class ClassDumper {
           // Skip unexpected files (such as a lock file) included in JAR file
           // https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1048 was caused
           // by an IOException
-          logger.warning("Corrupt jar file " + jar + "; could not load " + classFileName);
+          corruptedClassFileNames.add(classFileName);
         } else {
           throw new IOException(
               "Corrupt jar file " + jar + "; could not load " + classFileName, ex);
@@ -419,6 +422,14 @@ class ClassDumper {
             ex);
       }
     }
+
+    ImmutableList<String> corruptedFiles = corruptedClassFileNames.build();
+    int corruptedFileCount = corruptedFiles.size();
+    if (corruptedFileCount > 0) {
+      logger.warning("Corrupt jar file " + jar + "; could not load " + corruptedFiles.get(0)
+      + (corruptedFileCount > 1 ? " and other " + (corruptedFileCount -1) + " files" : ""));
+    }
+
     return javaClasses.build();
   }
 

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
@@ -35,6 +35,7 @@ import java.net.URLClassLoader;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.List;
+import java.util.logging.Logger;
 import javax.annotation.Nullable;
 import org.apache.bcel.Const;
 import org.apache.bcel.classfile.Attribute;
@@ -69,6 +70,7 @@ import org.apache.bcel.util.ClassPath;
  * in them, through the input class path for a linkage check.
  */
 class ClassDumper {
+  private static final Logger logger = Logger.getLogger(ClassDumper.class.getName());
 
   private final ImmutableList<Path> inputClassPath;
   private final FixedSizeClassPathRepository classRepository;
@@ -395,7 +397,16 @@ class ClassDumper {
         javaClasses.add(javaClass);
       } catch (ClassNotFoundException ex) {
         // We couldn't find the class in the jar file where we found it.
-        throw new IOException("Corrupt jar file " + jar + "; could not load " + classFileName, ex);
+        Throwable cause = ex.getCause();
+        if (cause != null && cause instanceof IOException) {
+          // Skip unexpected files (such as a lock file) included in JAR file
+          // https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1048 was caused
+          // by an IOException
+          logger.warning("Corrupt jar file " + jar + "; could not load " + classFileName);
+        } else {
+          throw new IOException(
+              "Corrupt jar file " + jar + "; could not load " + classFileName, ex);
+        }
       } catch (ClassFormatException ex) {
         // We couldn't load the class from the jar file where we found it.
         throw new IOException(

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
@@ -149,19 +149,24 @@ class ClassDumper {
   /**
    * Returns a map from classes to the symbol references they contain.
    */
-  SymbolReferenceMaps findSymbolReferences() throws IOException {
+  SymbolReferenceMaps findSymbolReferences() {
     SymbolReferenceMaps.Builder builder = new SymbolReferenceMaps.Builder();
 
     for (Path jar : inputClassPath) {
-      for (JavaClass javaClass : listClasses(jar)) {
-        if (!isCompatibleClassFileVersion(javaClass)) {
-          continue;
+      try {
+        for (JavaClass javaClass : listClasses(jar)) {
+          if (!isCompatibleClassFileVersion(javaClass)) {
+            continue;
+          }
+          String className = javaClass.getClassName();
+          // In listClasses(jar), ClassPathRepository creates JavaClass through the first JAR file
+          // that contains the class. It may be different from "jar" for an overlapping class.
+          ClassFile source = new ClassFile(findClassLocation(className), className);
+          builder.addAll(findSymbolReferences(source, javaClass));
         }
-        String className = javaClass.getClassName();
-        // In listClasses(jar), ClassPathRepository creates JavaClass through the first JAR file
-        // that contains the class. It may be different from "jar" for an overlapping class.
-        ClassFile source = new ClassFile(findClassLocation(className), className);
-        builder.addAll(findSymbolReferences(source, javaClass));
+      } catch (IOException ex) {
+        // For example, when a JAR file contains an unexpected lock file
+        logger.warning("Skipped JAR file " + jar + ": " + ex.getMessage());
       }
     }
 
@@ -385,9 +390,6 @@ class ClassDumper {
    */
   private ImmutableSet<JavaClass> listClasses(Path jar) throws IOException {
     ImmutableSet.Builder<JavaClass> javaClasses = ImmutableSet.builder();
-
-    ImmutableList.Builder<String> corruptedClassFileNames = ImmutableList.builder();
-
     for (String classFileName : listClassFileNames(jar)) {
       if (classFileName.startsWith("META-INF.versions.")) {
         // Linkage Checker does not support multi-release JAR (for Java 9+) yet
@@ -400,16 +402,7 @@ class ClassDumper {
         javaClasses.add(javaClass);
       } catch (ClassNotFoundException ex) {
         // We couldn't find the class in the jar file where we found it.
-        Throwable cause = ex.getCause();
-        if (cause != null && cause instanceof IOException) {
-          // Skip unexpected files (such as a lock file) included in JAR file
-          // https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1048 was caused
-          // by an IOException
-          corruptedClassFileNames.add(classFileName);
-        } else {
-          throw new IOException(
-              "Corrupt jar file " + jar + "; could not load " + classFileName, ex);
-        }
+        throw new IOException("Corrupt jar file " + jar + "; could not load " + classFileName, ex);
       } catch (ClassFormatException ex) {
         // We couldn't load the class from the jar file where we found it.
         throw new IOException(
@@ -422,14 +415,6 @@ class ClassDumper {
             ex);
       }
     }
-
-    ImmutableList<String> corruptedFiles = corruptedClassFileNames.build();
-    int corruptedFileCount = corruptedFiles.size();
-    if (corruptedFileCount > 0) {
-      logger.warning("Corrupt jar file " + jar + "; could not load " + corruptedFiles.get(0)
-      + (corruptedFileCount > 1 ? " and other " + (corruptedFileCount -1) + " files" : ""));
-    }
-
     return javaClasses.build();
   }
 

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
@@ -149,24 +149,19 @@ class ClassDumper {
   /**
    * Returns a map from classes to the symbol references they contain.
    */
-  SymbolReferenceMaps findSymbolReferences() {
+  SymbolReferenceMaps findSymbolReferences() throws IOException {
     SymbolReferenceMaps.Builder builder = new SymbolReferenceMaps.Builder();
 
     for (Path jar : inputClassPath) {
-      try {
-        for (JavaClass javaClass : listClasses(jar)) {
-          if (!isCompatibleClassFileVersion(javaClass)) {
-            continue;
-          }
-          String className = javaClass.getClassName();
-          // In listClasses(jar), ClassPathRepository creates JavaClass through the first JAR file
-          // that contains the class. It may be different from "jar" for an overlapping class.
-          ClassFile source = new ClassFile(findClassLocation(className), className);
-          builder.addAll(findSymbolReferences(source, javaClass));
+      for (JavaClass javaClass : listClasses(jar)) {
+        if (!isCompatibleClassFileVersion(javaClass)) {
+          continue;
         }
-      } catch (IOException ex) {
-        // For example, when a JAR file contains an unexpected lock file
-        logger.warning("Skipped JAR file " + jar + ": " + ex.getMessage());
+        String className = javaClass.getClassName();
+        // In listClasses(jar), ClassPathRepository creates JavaClass through the first JAR file
+        // that contains the class. It may be different from "jar" for an overlapping class.
+        ClassFile source = new ClassFile(findClassLocation(className), className);
+        builder.addAll(findSymbolReferences(source, javaClass));
       }
     }
 
@@ -390,6 +385,9 @@ class ClassDumper {
    */
   private ImmutableSet<JavaClass> listClasses(Path jar) throws IOException {
     ImmutableSet.Builder<JavaClass> javaClasses = ImmutableSet.builder();
+
+    ImmutableList.Builder<String> corruptedClassFileNames = ImmutableList.builder();
+
     for (String classFileName : listClassFileNames(jar)) {
       if (classFileName.startsWith("META-INF.versions.")) {
         // Linkage Checker does not support multi-release JAR (for Java 9+) yet
@@ -402,7 +400,7 @@ class ClassDumper {
         javaClasses.add(javaClass);
       } catch (ClassNotFoundException ex) {
         // We couldn't find the class in the jar file where we found it.
-        throw new IOException("Corrupt jar file " + jar + "; could not load " + classFileName, ex);
+        corruptedClassFileNames.add(classFileName);
       } catch (ClassFormatException ex) {
         // We couldn't load the class from the jar file where we found it.
         throw new IOException(
@@ -414,6 +412,19 @@ class ClassDumper {
                 + ex.getMessage(),
             ex);
       }
+    }
+
+    ImmutableList<String> corruptedFiles = corruptedClassFileNames.build();
+    int corruptedFileCount = corruptedFiles.size();
+    if (corruptedFileCount > 0) {
+      logger.warning(
+          "Corrupt files in "
+              + jar
+              + "; could not load "
+              + corruptedFiles.get(0)
+              + (corruptedFileCount > 1
+                  ? " and other " + (corruptedFileCount - 1) + " files"
+                  : ""));
     }
     return javaClasses.build();
   }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
@@ -401,16 +401,6 @@ class ClassDumper {
       } catch (ClassNotFoundException ex) {
         // We couldn't find the class in the jar file where we found it.
         corruptedClassFileNames.add(classFileName);
-      } catch (ClassFormatException ex) {
-        // We couldn't load the class from the jar file where we found it.
-        throw new IOException(
-            "Corrupt jar file "
-                + jar
-                + "; could not load "
-                + classFileName
-                + "; "
-                + ex.getMessage(),
-            ex);
       }
     }
 

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java
@@ -375,9 +375,17 @@ public class ClassDumperTest {
     Artifact kinesisClient = new DefaultArtifact("com.amazonaws:amazon-kinesis-client:1.13.0");
     List<Path> paths = ClassPathBuilder.artifactsToClasspath(ImmutableList.of(kinesisClient));
     ClassDumper classDumper = ClassDumper.create(paths);
+    Path kinesisJar = paths.get(0);
 
     // This should not raise IOException
     SymbolReferenceMaps symbolReferences = classDumper.findSymbolReferences();
     assertNotNull(symbolReferences);
+
+    Truth.assertWithMessage("Invalid files should not stop loading valid class files")
+        .that(symbolReferences.getClassToClassSymbols().keySet())
+        .comparingElementsUsing(
+            Correspondence.transforming(
+                (ClassFile classFile) -> classFile.getJar(), "is in the JAR file"))
+        .contains(kinesisJar);
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java
@@ -17,6 +17,7 @@
 package com.google.cloud.tools.opensource.classpath;
 
 import static com.google.cloud.tools.opensource.classpath.TestHelper.absolutePathOfResource;
+import static org.junit.Assert.assertNotNull;
 
 import com.google.common.base.VerifyException;
 import com.google.common.collect.ImmutableList;
@@ -313,7 +314,8 @@ public class ClassDumperTest {
             ImmutableList.of(absolutePathOfResource("testdata/conscrypt-openjdk-uber-1.4.2.jar")));
 
     try {
-      classDumper.isUnusedClassSymbolReference("org.conscrypt.Conscrypt", new ClassSymbol("dummy.NoSuchClass"));
+      classDumper.isUnusedClassSymbolReference(
+          "org.conscrypt.Conscrypt", new ClassSymbol("dummy.NoSuchClass"));
 
       Assert.fail("It should throw VerifyException when it cannot find a class symbol reference");
     } catch (VerifyException ex) {
@@ -363,5 +365,19 @@ public class ClassDumperTest {
         .doesNotContain(
             new ClassFile(
                 sisuGuicePath, "com.google.inject.internal.InjectorImpl$BindingsMultimap"));
+  }
+
+  @Test
+  public void testListClasses_unexpectedNonClassFile() throws RepositoryException, IOException {
+    // com.amazonaws:amazon-kinesis-client:1.13.0 contains an unexpected lock file
+    // /unison/com/e007f77498fd27177e2ea931a06dcf50/unison/tmp/amazonaws/services/kinesis/leases/impl/LeaseTaker.class
+    // https://github.com/awslabs/amazon-kinesis-client/issues/654
+    Artifact kinesisClient = new DefaultArtifact("com.amazonaws:amazon-kinesis-client:1.13.0");
+    List<Path> paths = ClassPathBuilder.artifactsToClasspath(ImmutableList.of(kinesisClient));
+    ClassDumper classDumper = ClassDumper.create(paths);
+
+    // This should not raise IOException
+    SymbolReferenceMaps symbolReferences = classDumper.findSymbolReferences();
+    assertNotNull(symbolReferences);
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java
@@ -379,7 +379,6 @@ public class ClassDumperTest {
 
     // This should not raise IOException
     SymbolReferenceMaps symbolReferences = classDumper.findSymbolReferences();
-    assertNotNull(symbolReferences);
 
     Truth.assertWithMessage("Invalid files should not stop loading valid class files")
         .that(symbolReferences.getClassToClassSymbols().keySet())

